### ISSUE TITLE
[BUG] When AppCenter directory is not in "Assets" Unity can't build

### ISSUE THREAD

**Issue by liszto** [Microsoft/AppCenter-SDK-Unity#321]
Hello devs ! 

In our pipeline, every plugin must be in Assets/Plugins. So we have to change a little bit some plugin hierarchy (yours for example). Every time we do this, it's ok or barely ok. But with your plugin, it requires to be in "Assets/AppCenter" cause **if the folder doesn't exist, it can't generate AppCenterSettings.asset** and throw this error : 
`Player build failed: Parent directory must exist before creating asset at Assets/AppCenter/AppCenterSettings.asset.`

A cleaner way to do this, would be to generate it locally to your AppCenter folder location instead of a static path named "Assets/AppCenter". 

Thanks in advance for your answers.

**Comment by liszto** [comment]
Or currently you can do this as a fast fix in AppCenterSettingsContext.cs : 

```
public const string AppCenterParentFolder = "Assets";
public const string AppCenterFolder = "AppCenter";
private const string SettingsPath = AppCenterParentFolder + "/" + AppCenterFolder+ "/AppCenterSettings.asset";
private const string AdvancedSettingsPath = AppCenterParentFolder + "/" + AppCenterFolder + "/AppCenterSettingsAdvanced.asset";

    public static AppCenterSettings SettingsInstance
    {
        get
        {
            // No need to lock because this can only be accessed from the main thread.
            var instance = AssetDatabase.LoadAssetAtPath<AppCenterSettings>(SettingsPath);
            if (instance == null)
            {
				if(!AssetDatabase.IsValidFolder(AppCenterParentFolder + "/" + AppCenterFolder))
				{
					AssetDatabase.CreateFolder(AppCenterParentFolder, AppCenterFolder);
				}

				instance = CreateInstance<AppCenterSettings>();
                AssetDatabase.CreateAsset(instance, SettingsPath);
                AssetDatabase.SaveAssets();
            }
            return instance;
        }
    }
```
In this way, if the folder doesn't exist, it will create it and the build will succeed. But the file generation is not made where the plugin really is.

P.S : I renamed `AppCenterPath` by `AppCenterParentFolder` which is much logical when I see what it is :)

**Comment by annakocheshkova** [comment]
Hi @liszto! We support other directories than default one, just it's not documented. 
You were going in the right direction. Please refer to [this comment](https://github.com/Microsoft/AppCenter-SDK-Unity/issues/115#issuecomment-427024648).

**Comment by liszto** [comment]
Indeed I saw this with my lil code update just above.
But due to some hierarchy modification just change the first path won't solve entierly my problem. There is others location in your code base where the global path will be wrong due to this : 
`Assets/AppCenter/Plugins/WhateverFolderPath`
because in our complete project logic it will be : 
`Assets/Plugins/AppCenter/WhateverFolderPath`.

But for this, it will probably be a patch that we will need to do each time we will update your plugin. Or we/you could maybe use some tricks to retrieve this (it's editor only from what I remember so it won't be a huge cost issue).

When I need to do this, I retrieve the current script path with reflection. Then I will keep only the interesting path part. Only the path part after "Assets/" and then just generate each path in your code based on this.

For example in `AppCenterSettingsMakerAndroid` instead of having : 

```
 public bool IsAnalyticsAvailable()
    {
        return File.Exists(AppCenterSettingsContext.AppCenterPath + "AppCenter/Plugins/Android/appcenter-analytics-release.aar");
    }
```

I will retrieve the current script path by doing : 

```
string GetCurrentFileName([System.Runtime.CompilerServices.CallerFilePath] string fileName = null)
{
         return fileName;
}
```

Then I call this method into `AppCenterSettingsMakerAndroid` and this will return me : 
`D:\Project\MyHemisForFlexom\UnityProject\Assets\Plugins\AppCenter\Scripts\Editor\AppCenterSettingsMakerAndroid.cs`

And with this, I will split the string with the "Assets\" or "Assets/" depending of the platform separator that could be retrieve with `Path.DirectorySeparatorChar` that return `\` on Windows system and `/` on Unix system.

Then I can use this : 

```
 public bool IsAnalyticsAvailable()
    {
        return File.Exists( UnityInternalAppCenterPath + "/Android/appcenter-analytics-release.aar");
    }
```

**Where `UnityInternalAppCenterPath ` value will be : Assets/Plugins/AppCenter. In this way you will always have the entire path from Assets folder to AppCenter folder** without editing a script or a scriptable each time. I would keep the override possibility from your scriptable object just in case a user would modify the hierarchy deeper than this. For example if the user would replace (for any reason) the "AppCenter" word by "MyOwnStuff" or anything else.

Sorry for this huge post :( quite hard to explain entierely and "clearly" what I do to solve that kind of potential trouble.

**Comment by annakocheshkova** [comment]
I see @liszto, that's quite an interesting suggestion. When we were implementing this, we were also looking for the ways to safely retrieve this path but somehow we stopped at the easy solution at that time. 
But still I don't quite get why simply setting `AppCenterPath` to `Assets/Plugins` don't work in your case?

**Comment by liszto** [comment]
Because in the current state, I have to edit your code to make it work. 
My main concern isto not modify the codebase as far as I can.

Currently your code is : 
AppCenterPath + "AppCenter/Plugins"
whereas in our configuration it's : 
AppCenter + "Plugins/AppCenter" 

So in the current state it's not possible without editing your code (What I'm currently doing)

**Comment by annakocheshkova** [comment]
Ok I understand, it makes sense. I will create bug for this.

**Comment by liszto** [comment]
Thanks really !

**Comment by AnastasiaKubova** [comment]
The fix had been released in [2.6.0](https://github.com/microsoft/appcenter-sdk-unity/releases/tag/2.6.0).
